protocols/kad/src/behaviour.rs

```diff
@@ -1045,6 +1045,11 @@ where
                         },
                     }),
             );
+
+        self.queued_events
+            .push_back(ToSwarm::GenerateEvent(KademliaEvent::ModeChanged {
+                mode: self.mode,
+            }));
```

Comment:
This is also called when we set the mode manually, see line 1014. I think we should move this to `determine_mode_from_external_addresses`.

Reply:
We'll need to slightly modify that function though because at the moment, we don't know whether the mode actually changed at the end. We'll need to perform some diffing there and only emit the event if we actually switched from client to server or vice-versa.

Reply:
done